mgosimplekv/kv.go

```diff
@@ -0,0 +1,166 @@
+type kvStore struct {
+	coll *mgo.Collection
+}
+
+// NewStore returns a new Store implementation that uses
+// the given mongo collection for storage.
+func NewStore(coll *mgo.Collection) (simplekv.Store, error) {
+	if err := coll.EnsureIndex(mgo.Index{
+		Key:         []string{"expire"},
+		ExpireAfter: time.Second,
+	}); err != nil {
+		return nil, errgo.Mask(err)
+	}
+	return &kvStore{
+		coll: coll,
+	}, nil
+}
+
+type sessionKey struct{}
+
+// Context implements simplekv.Context by copying the kvStore's underlying
+// session if one isn't already present in the context.
+func (s *kvStore) Context(ctx context.Context) (_ context.Context, close func()) {
+	if session, _ := ctx.Value(sessionKey{}).(*mgo.Session); session != nil {
+		return ctx, func() {}
+	}
+	// TODO (mhilton) add some more advanced session pooling.
+	session := s.coll.Database.Session.Copy()
+	return context.WithValue(ctx, sessionKey{}, session), session.Close
+}
+
+// s returns a *mgo.Session for use in subsequent queries. The returned
+// session must be closed once finished with.
+func (s *kvStore) session(ctx context.Context) *mgo.Session {
+	if s, _ := ctx.Value(sessionKey{}).(*mgo.Session); s != nil {
+		return s.Clone()
+	}
+	return s.coll.Database.Session.Copy()
+}
+
+// c returns a *mgo.Collection with the given name in the current
+// database. The collection's underlying session must be closed when the
+// query is complete.
+func (s *kvStore) c(ctx context.Context) *mgo.Collection {
+	return s.coll.With(s.session(ctx))
+}
+
+type kvDoc struct {
+	Key    string    `bson:"_id"`
+	Value  []byte    `bson:"value'`
+	Expire time.Time `bson:",omitempty"`
+}
+
+// Get implements simplekv.Store.Get by retrieving the document with
+// the given key from the store's collection.
+func (s *kvStore) Get(ctx context.Context, key string) ([]byte, error) {
+	coll := s.c(ctx)
+	defer coll.Database.Session.Close()
+
+	var doc kvDoc
+	if err := coll.FindId(key).One(&doc); err != nil {
+		if errgo.Cause(err) == mgo.ErrNotFound {
+			return nil, simplekv.KeyNotFoundError(key)
+		}
+		return nil, errgo.Mask(err)
+	}
+	return doc.Value, nil
+}
+
+// Set implements simplekv.Store.Set by upserting the document with
+// the given key, value and expire time into the store's collection.
+func (s *kvStore) Set(ctx context.Context, key string, value []byte, expire time.Time) error {
+	coll := s.c(ctx)
+	defer coll.Database.Session.Close()
+
+	_, err := coll.UpsertId(key, bson.D{{
+		"$set", bson.D{{
+			"value", value,
+		}, {
+			"expire", expire,
+		}},
+	}})
+	return errgo.Mask(err)
+}
+
+var updateStrategy = retry.Exponential{
+	Initial:  time.Microsecond,
+	Factor:   2,
+	MaxDelay: 500 * time.Millisecond,
+	Jitter:   true,
+}
+
+func (s *kvStore) Update(ctx context.Context, key string, expire time.Time, getVal func(old []byte) ([]byte, error)) error {
+	coll := s.c(ctx)
+	defer coll.Database.Session.Close()
+
+	for r := retry.Start(updateStrategy, nil); r.Next(); {
```

Comment:
As this is looping you should probably be checking that the context hasn't been closed somewhere in the loop. if it has then the appropriate error should be returned unmasked.
Ideally retry would support being interrupted by a context being closed, but that may be a little too much.